### PR TITLE
adds capability to resolve anchors and aliases in yaml in `aws cloudformation package`

### PR DESCRIPTION
**Pull request by shortjared**
*Issue #, if available:*

No issue number specifically, but this has been mentioned before in several cases and a feature folks have wanted natively in CloudFormation but we aren't going to get anytime soon.

*Description of changes:*

This simply enhances the yaml parser used to resolve anchors and aliases to a non-referential structure. This lets you do a little bit DRYing up your templates without any external tooling. This should be fully backwards compatible and you only need to use it if you want to.

Important to note, this is doing _nothing_ that nearly all other yaml libs can't do. Take the below template for example and plug it in to http://yaml-online-parser.appspot.com and you'll get perfectly valid json outputs with no additional logic. Of course, you would still need `aws cloudformation package` to populate the `CodeUri` but that is a different process.

Also note, the usage of the Metadata section is inconsequential. It's just an easy place to put stuff without impacting the template. You could just as well anchor to the first instance of properties on your actual resource instance.

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: 'AWS::Serverless-2016-10-31'
Description: A starter AWS Lambda function.

Metadata:
  Templates:
    Resources:
      Properties: &Props
        Handler: lambda_function.lambda_handler
        Runtime: python3.6
        CodeUri: .
        Description: A starter AWS Lambda function.
        MemorySize: 128
        Timeout: 3

Resources:
  helloworldpython3:
    Type: 'AWS::Serverless::Function'
    Properties:
      <<: *Props

  helloworldnode:
    Type: 'AWS::Serverless::Function'
    Properties:
      <<: *Props
      Runtime: nodejs8.10
```

Example output from `aws cloudformation package` in this case would be

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Description: A starter AWS Lambda function.
Metadata:
  Templates:
    Resources:
      Properties:
        CodeUri: .
        Description: A starter AWS Lambda function.
        Handler: lambda_function.lambda_handler
        MemorySize: 128
        Runtime: python3.6
        Timeout: 3
Resources:
  helloworldnode:
    Properties:
      CodeUri: s3://serverless-deployments-us-east-1-123456789101/ec854419bcd553ab2fa06d17aa10f7c9
      Description: A starter AWS Lambda function.
      Handler: lambda_function.lambda_handler
      MemorySize: 128
      Runtime: nodejs8.10
      Timeout: 3
    Type: AWS::Serverless::Function
  helloworldpython3:
    Properties:
      CodeUri: s3://serverless-deployments-us-east-1-123456789101/ec854419bcd553ab2fa06d17aa10f7c9
      Description: A starter AWS Lambda function.
      Handler: lambda_function.lambda_handler
      MemorySize: 128
      Runtime: python3.6
      Timeout: 3
    Type: AWS::Serverless::Function
Transform: AWS::Serverless-2016-10-31
```

### CODE REVIEW

**Comment by jfuss** [comment]
I have commented on AWS SAM CLI's PR that was linked to this but the reasons hold true here as well. 

You can see detailed comments on why the PR was not accepted [here](https://github.com/awslabs/aws-sam-cli/pull/566#issuecomment-425171106).

TLDR:
Their is a requirement that a template with the Transform of AWS::Serverless-2016-10-31 needs to be deployable through CloudFormation. This change will break that requirement.

**Comment by mswilson** [comment]
Obsoleted by https://github.com/aws/aws-cli/pull/3827 ?
